homeassistant/auth/models.py

```diff
@@ -7,22 +7,49 @@
 
 from homeassistant.util import dt as dt_util
 
+from . import permissions
 from .util import generate_secret
 
 TOKEN_TYPE_NORMAL = 'normal'
 TOKEN_TYPE_SYSTEM = 'system'
 TOKEN_TYPE_LONG_LIVED_ACCESS_TOKEN = 'long_lived_access_token'
 
 
+@attr.s(slots=True)
+class Group:
+    """A group."""
+
+    name = attr.ib(type=str)  # type: Optional[str]
+    policy = attr.ib(type=Dict[str, dict])
+    id = attr.ib(type=str, default=attr.Factory(lambda: uuid.uuid4().hex))
+    # System generated groups cannot be changed
+    system_generated = attr.ib(type=bool, default=False)
+
+    _permissions = attr.ib(
+        type=permissions.PolicyPermissions,
+        init=False,
+        cmp=False,
+        default=None,
+    )
+
+    @property
+    def permissions(self):
+        """Return group permissions."""
+        if self._permissions is None:
+            self._permissions = permissions.PolicyPermissions(self.policy)
+
+        return self._permissions
+
+
 @attr.s(slots=True)
 class User:
     """A user."""
 
     name = attr.ib(type=str)  # type: Optional[str]
+    group = attr.ib(type=Group)
```

Comment:
User only belongs to one group?

Reply:
Yep. Since policies are attached to groups, having multiple groups becomes madness

Reply:
And with madness I mean it becomes very expensive to verify each entity access.

Reply:
I'll open an architecture issue.

Reply:
btw, the group code is in this PR but that's PR #16935. It was just needed for this PR to progress.